daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs

```diff
@@ -286,19 +289,39 @@ recoverPackageName pkg pid= do
     Nothing -> LF.unPackageId pid
 
 -- | Download all Packages reachable from a PackageId; fail if any don't exist or can't be decoded.
-downloadAllReachablePackages :: LedgerArgs -> [LF.PackageId] -> IO (Map LF.PackageId LF.Package)
-downloadAllReachablePackages args pids = loop Map.empty (Set.fromList pids)
+downloadAllReachablePackages ::
+       (LF.PackageId -> IO LF.Package)
+    -- ^ An IO action to download a package.
+    -> [LF.PackageId]
+    -- ^ The roots of the dependency tree we want to download.
+    -> [LF.PackageId]
+    -- ^ Exclude these package ids from downloading, because they are already present.
+    -> IO (Map LF.PackageId (Maybe LF.Package))
+    -- ^ Return a map of dependency package ids and maybe the downloaded package or Nothing, if the
+    -- package is needed but already present.
+downloadAllReachablePackages downloadPkg pids exclPids =
+    loop Map.empty (Set.fromList pids) (Set.fromList exclPids)
   where
-    loop :: Map LF.PackageId LF.Package -> Set.Set LF.PackageId -> IO (Map LF.PackageId LF.Package)
-    loop acc pkgIds =
+    loop ::
+           Map LF.PackageId (Maybe LF.Package)
+        -> Set.Set LF.PackageId
+        -> Set.Set LF.PackageId
+        -> IO (Map LF.PackageId (Maybe LF.Package))
+    loop acc pkgIds exclPids = do
         case Set.minView pkgIds of
             Nothing -> return acc
             Just (pid, morePids) ->
                 if pid `Map.member` acc
-                    then loop acc morePids
+                    then loop acc morePids exclPids
```

Comment:
Where do we change `exclPids`? It looks like it’s the unmodified in all recursive calls.

Reply:
The don't change, these are the package ids that are already present in the dependency database and hence don't need fetching. We already make sure that we don't fetch package ids that we already fetched in the loop.

Reply:
If they don’t change, can we stop passing them along in `loop`?

Reply:
yep, changed.